### PR TITLE
Behaviour as runtime dependency

### PR DESCRIPTION
**Pull request by marcandre**
With this PR, `@behaviour` declarations introduce only a runtime dependency instead of a compile-time.

The checks are moved from the module compiler to the parallel checker.

This requires storing the `@impl` statements as well as the list of `@behaviour`s when compiling the module, for later retrieval when doing the checks.

~Remains to be done: `@behaviour :"Elixir.MyModule"` does not (yet) introduce a runtime dependency at all and so it is possible that further changes to `MyModule` will not trigger checks for modules implementing it.~

Notes:
* I isolated the `behaviour` code to a new `Module.Behaviour` module, but that's not necessary and code could easily be moved back into `Module`

### CODE REVIEW

**Comment by marcandre** [comment]
PR ready for review. I presume the Windows OTP-25 failure is not related

File: lib/elixir/lib/module.ex [hunk 1]
```diff
@@ -1824,22 +1802,11 @@ defmodule Module do
   defp args_count([], total, defaults), do: {total, defaults}
 
   @doc false
-  def check_derive_behaviours_and_impls(env, set, bag, all_definitions) do
+  def check_derive_behaviours_and_impls(env, set, bag) do
```
Comment (josevalim) [inline lib/elixir/lib/module.ex:1805]: Shall we rename it to:

```suggestion
  def __check_attributes__(env, set, bag) do
```

As it is now mostly about validating certain attributes were set?

File: lib/elixir/lib/module.ex [hunk 2]
```diff
@@ -1824,22 +1802,11 @@ defmodule Module do
...
-
-    pending_callbacks =
-      if impls != [] do
-        {non_implemented_callbacks, contexts} = check_impls(env, behaviours, callbacks, impls)
-        warn_missing_impls(env, non_implemented_callbacks, contexts, all_definitions)
-        non_implemented_callbacks
-      else
-        callbacks
-      end
+    Module.Behaviour.force_runtime_dependencies(behaviours, env)
```
Comment (josevalim) [inline lib/elixir/lib/module.ex:1808]: I would move only the logic of this particular function in particular here. Everything else in Module.Behaviour is 💯 .

File: lib/elixir/lib/module/behaviour.ex [hunk 3]
```diff
@@ -0,0 +1,419 @@
+defmodule Module.Behaviour do
```
Comment (josevalim) [inline lib/elixir/lib/module/behaviour.ex:1]: Nitpick: let's please call this `Module.Types.Behaviour`.

**Review by josevalim — COMMENTED** [review]
This is fantastic @marcandre, thank you! :heart: I have dropped 3 comments but they are all nitpicks about renaming/moving code around. :)

File: lib/elixir/lib/module.ex (hunk 2 above)
Comment (marcandre) [reply lib/elixir/lib/module.ex:1808]: Yeah, I hesitated. Will move here 👍

**Comment by marcandre** [comment]
Great, thank you 😄
I should be able to update this today or tomorrow at the latest.

**Comment by marcandre** [comment]
@josevalim tweaks done

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**